File: 0383-ransom-note/solution.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        int n = ransomNote.length();
        int m = magazine.length();
        int[] freq = new int[26];

        for(int i=0; i<m; i++) {
            char ch1 = magazine.charAt(i);
            freq[ch1 - 'a']++;
        }

        for(int i=0; i<n; i++) {
            char ch2 = ransomNote.charAt(i);
            if(freq[ch2 - 'a'] == 0) return false;
            freq[ch2 - 'a']--;
        }
        return true;
    }
}

File: 0383-ransom-note/solution.swift
class Solution {
    func canConstruct(_ ransomNote: String, _ magazine: String) -> Bool {
        // count magazine characters
        var map1 = [Character: Int]()
        //{b-1}
        for ch in Array(magazine) {
            map1[ch, default: 0] += 1
        }

        var map2 = [Character: Int]()
        //{a-2}
        for ch in Array(ransomNote) {
            map2[ch, default: 0] += 1
        }

        for ch in Array(ransomNote) {
            if let j = map1[ch] {
                if map1[ch] == 1 {
                    map1.removeValue(forKey: ch)
                } else {
                    map1[ch]! -= 1
                }
            } else {
                return false
            }
        }
        
        return true
    }
}
